### PR TITLE
Romy - splitting the pr, service->

### PR DESCRIPTION
**Pull request by bossyshrimp68**
find good shooting pose, pt.2 will soon follow🔥

### CODE REVIEW

File: src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingPositionConstants.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package edu.greenblitz.robotName.shootingStateService;
+
+
+import edu.greenblitz.robotName.FieldConstants;
+import edu.greenblitz.robotName.utils.shootingCalculations.ShootingZone;
+import edu.wpi.first.math.geometry.Translation2d;
+
+public class ShootingPositionConstants {
+    public static final Translation2d CENTER_OF_CIRCLE = new Translation2d(1, 1);
+    public static final double CIRCLE_RADIUS_IN_METERS = 1;
```
Comment (YoniKiriaty) [inline src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingPositionConstants.java:10]: what circle?
specify which circle

File: src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingZone.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import edu.wpi.first.math.Pair;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Translation2d;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ShootingZone extends GBMath.GBCircle {
+
+    private List<Pair<Translation2d, Translation2d>> restrictedBounds; //first min, second max
```
Comment (YoniKiriaty) [inline src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingZone.java:14]:
```suggestion
    private List<Pair<Translation2d, Translation2d>> restrictedBounds; //First input is the lower bound, second input is the upper bound.
```

**Review by YoniKiriaty — CHANGES_REQUESTED** [review]

**Review by YoniKiriaty — APPROVED** [review]
Yoni Approved 😁👍👍

File: src/main/java/edu/greenblitz/robotName/utils/GBMath.java [hunk 3]
```diff
@@ -1,22 +1,159 @@
...
+     */
+    public static double distance(Translation3d p1, Translation3d p2) {
+        return distance(p1.getX(), p1.getY(), p1.getZ(), p2.getX(), p2.getY(), p2.getZ());
+    }
+
+    /**
+     * A circle.
+     * A radius and center position, EPSILON for when the given point is right below the center (Cx == Px)
+     */
+    public static class GBCircle {
```
Comment (YHerm) [inline src/main/java/edu/greenblitz/robotName/utils/GBMath.java:74]: I would not put a class in GBMath, maybe create package or get it a special file for it🤩

File: src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShooterAngle.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import edu.greenblitz.robotName.utils.GBMath;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.geometry.Translation3d;
+
+public class ShooterAngle {
+
+    public static Rotation2d getShooterAngleBasedOnPosition(Translation3d position) {
+        Translation2d relativeSpeakerPosition = getRelativePositionOfMiddleOfSpeaker(position);
+        double angle = Math.atan(relativeSpeakerPosition.getY() / relativeSpeakerPosition.getX());
```
Comment (YHerm) [inline src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShooterAngle.java:14]: consider changing angle to a more informative name, but this is not a must

File: src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShooterAngle.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        return new Rotation2d(angle);
+    }
+
+    public static Translation2d getRelativePositionOfMiddleOfSpeaker(Translation3d position) {
```
Comment (YHerm) [inline src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShooterAngle.java:18]: try to find a better name maybe

File: src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShooterAngle.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        Translation3d Speaker = FieldConstants.MID_SPEAKER_POSITION;
+        double x = GBMath.distance(position, new Translation3d(Speaker.getX(), Speaker.getY(), 0));
+        double y = Speaker.getZ() - position.getZ();
```
Comment (YHerm) [inline src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShooterAngle.java:21]: like in angle, consider more informative name

File: src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingZone.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+            Translation2d upperLimit = bound.getSecond();
+            Translation2d lowerLimit = bound.getFirst();
+            if (targetPosition.getY() > lowerLimit.getY() && targetPosition.getY() < upperLimit.getY()) {
+                double upperLimitTargetDelta = upperLimit.getY() - targetPosition.getY();
+                double lowerLimitTargetDelta = targetPosition.getY() - lowerLimit.getY();
+                return upperLimitTargetDelta > lowerLimitTargetDelta ? lowerLimit : upperLimit;
+            }
+        }
+        return targetPosition;
+    }
```
Comment (YHerm) [inline src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingZone.java:47]: Consider to split the function

File: src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingZone.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+import edu.wpi.first.math.Pair;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Translation2d;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ShootingZone extends GBMath.GBCircle {
+
+    private List<Pair<Translation2d, Translation2d>> restrictedBounds; //first is the lower bound, second is the upper
```
Comment (YHerm) [inline src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingZone.java:15]: you can maybe remove the comment and in the build function do the finding min max🛠️

File: src/main/java/edu/greenblitz/robotName/FieldConstants.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.List;
+
+/**
+ * All constants in this file are measured in meters.
+ * <p>
+ * When talking about 3d objects the z direction is up.
+ */
+public class FieldConstants {
+
+    public static final Translation3d MID_SPEAKER_POSITION = new Translation3d(2.225, 0.23, 2.045);
```
Comment (YHerm) [inline src/main/java/edu/greenblitz/robotName/FieldConstants.java:18]: What is MID speaker?

File: src/main/java/edu/greenblitz/robotName/utils/GBMath.java [hunk 10]
```diff
@@ -1,22 +1,159 @@
...
+         * @return The position of the closest point.
+         */
+        public Translation2d getClosestPositionOnBorder(Translation2d position) {
+            if (position.getX() == centerPosition.getX()) {
+                return position.plus(new Translation2d(EPSILON, 0));
+            }
+            double slope = ((centerPosition.getY() - position.getY()) / centerPosition.getX() - position.getX());
+            double angle = Math.atan(slope);
+            double x = radius * Math.cos(angle) * Math.signum(position.getX() - centerPosition.getX()) + centerPosition.getX();
+            double y = radius * Math.sin(angle) * Math.signum(position.getX() - centerPosition.getX()) + centerPosition.getY();
```
Comment (YHerm) [inline src/main/java/edu/greenblitz/robotName/utils/GBMath.java:155]: consider to use more informative name🏟️

File: src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingZone.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    public void addBound(Pair<Translation2d, Translation2d> bound) {
+        restrictedBounds.add(bound);
+    }
+
+    public void setRestrictedBounds(List<Pair<Translation2d, Translation2d>> restrictedBounds) {
+        this.restrictedBounds = restrictedBounds;
+    }
+
+    @Override
+    public Translation2d getClosestPositionOnBorder(Translation2d position) {
```
Comment (YHerm) [inline src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingZone.java:35]: what border?

File: src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingState.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.geometry.Translation3d;
+import edu.greenblitz.robotName.shootingStateService.ShootingPositionConstants.*;
+
+import static edu.greenblitz.robotName.shootingStateService.ShootingPositionConstants.LEGAL_SHOOTING_ZONE;
+
+public class ShootingState {
+
+    public static boolean isRobotPositionFine() {
```
Comment (tnrz135) [inline src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingState.java:17]: What is fine? How about isRobotInShootingPosition?

File: src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShooterAngle.java (hunk 4 above)
Comment (tnrz135) [reply src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShooterAngle.java:14]: It is a must now

**Review by tnrz135 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingState.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+public class ShootingState {
+
+    public static boolean isRobotInShootingPosition() {
+        Pose2d robotPose = SwerveChassis.getInstance().getRobotPose();
+        return LEGAL_SHOOTING_ZONE.isInCircle(robotPose.getTranslation());
+    }
+
+    public static Translation2d getRobotTargetTranslation() {
+        Pose2d robotPose = SwerveChassis.getInstance().getRobotPose();
+        if (LEGAL_SHOOTING_ZONE.isInCircle(robotPose.getTranslation())) {
```
Comment (itamaroryan) [inline src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingState.java:23]: Duplicate code 😠😠😠. Change to `isRobotInShootingPosition()`.

**Review by itamaroryan — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingAngle.java [hunk 14]
```diff
@@ -0,0 +1,24 @@
+package edu.greenblitz.robotName.utils.shootingCalculations;
+
+
+import edu.greenblitz.robotName.FieldConstants;
+import edu.greenblitz.robotName.utils.GBMath;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.geometry.Translation3d;
+
+public class ShootingAngle {
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingAngle.java:10]: is it a good name for the class?

File: src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingAngle.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+    public static Rotation2d getShootingAngleBasedOnPosition(Translation3d position) {
+        Translation2d relativeSpeakerPosition = getRelativePositionToMiddleOfSpeaker(position);
+        double targetAngle = Math.atan(relativeSpeakerPosition.getY() / relativeSpeakerPosition.getX());
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingAngle.java:14]: atan2?

File: src/main/java/edu/greenblitz/robotName/utils/GBCircle.java
Comment (yonnbro) [inline src/main/java/edu/greenblitz/robotName/utils/GBCircle.java:1]: this is butiful, באמת מתועד מצוין

File: src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingZone.java [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+			Translation2d upperLimit = bound.getSecond();
+			Translation2d lowerLimit = bound.getFirst();
+			if (targetPosition.getY() > lowerLimit.getY() && targetPosition.getY() < upperLimit.getY()) {
+				return getClosestLimitToPosition(upperLimit, lowerLimit, targetPosition);
+			}
+		}
+		return targetPosition;
+	}
+
+	public Rotation2d getTargetRobotAngle(Translation2d position) {
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/utils/shootingCalculations/ShootingZone.java:55]:
```suggestion
	public Rotation2d getTargetAngleForRobot(Translation2d position) {
```

or sometyhing

**Review by NoamRosenberg08 — CHANGES_REQUESTED** [review]

**Review by yonnbro — APPROVED** [review]

**Review by NoamRosenberg08 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**